### PR TITLE
FEAT: Lora multi support

### PR DESCRIPTION
**Pull request by hainaweiben**

### CODE REVIEW

File: xinference/api/restful_api.py [hunk 1]
```diff
@@ -50,6 +50,8 @@
 from uvicorn import Config, Server
 from xoscar.utils import get_next_port
 
+from xinference.utils import PeftModelConfig
```
Comment (ChengjieLi28) [inline xinference/api/restful_api.py:53]: Use relative import

File: xinference/api/restful_api.py [hunk 2]
```diff
@@ -692,9 +694,7 @@ async def launch_model(
         replica = payload.get("replica", 1)
         n_gpu = payload.get("n_gpu", "auto")
         request_limits = payload.get("request_limits", None)
-        peft_model_path = payload.get("peft_model_path", None)
-        image_lora_load_kwargs = payload.get("image_lora_load_kwargs", None)
-        image_lora_fuse_kwargs = payload.get("image_lora_fuse_kwargs", None)
+        peft_model_config_dict = payload.get("peft_model_config", None)
```
Comment (ChengjieLi28) [inline xinference/api/restful_api.py:697]: Just naming this as `peft_model_config`

File: xinference/utils.py [hunk 3]
```diff
@@ -12,9 +12,38 @@
...
+
 import torch
 
 
 def cuda_count():
     # even if install torch cpu, this interface would return 0.
     return torch.cuda.device_count()
+
+
+class PeftModelConfig:
```
Comment (ChengjieLi28) [inline xinference/utils.py:25]: Place this class to `xinference/types.py`

File: xinference/client/restful/restful_client.py [hunk 4]
```diff
@@ -18,6 +18,8 @@
 
 import requests
 
+from xinference.utils import PeftModelConfig
```
Comment (ChengjieLi28) [inline xinference/client/restful/restful_client.py:21]: Ditto

File: xinference/client/restful/restful_client.py [hunk 5]
```diff
@@ -829,15 +829,11 @@ def launch_model(
         n_gpu: Optional[Union[int, str]],
             The number of GPUs used by the model, default is "auto".
             ``n_gpu=None`` means cpu only, ``n_gpu=auto`` lets the system automatically determine the best number of GPUs to use.
+        mode_deft_config: Optional[PeftModelConfig],
```
Comment (ChengjieLi28) [inline xinference/client/restful/restful_client.py:832]: Typo. `deft`?

File: xinference/client/restful/restful_client.py [hunk 6]
```diff
@@ -829,15 +829,11 @@ def launch_model(
...
+            The config for the lora model.
```
Comment (ChengjieLi28) [inline xinference/client/restful/restful_client.py:833]: A more detailed description of how this data structure should be set up in the case of multi lora

File: xinference/core/supervisor.py [hunk 7]
```diff
@@ -22,6 +22,8 @@
 
 import xoscar as xo
 
+from xinference.utils import PeftModelConfig
```
Comment (ChengjieLi28) [inline xinference/core/supervisor.py:25]: Ditto

File: xinference/deploy/cmdline.py [hunk 8]
```diff
@@ -22,6 +22,8 @@
 import click
 from xoscar.utils import get_next_port
 
+from xinference.utils import PeftModelConfig
```
Comment (ChengjieLi28) [inline xinference/deploy/cmdline.py:25]: Ditto

File: xinference/deploy/cmdline.py [hunk 9]
```diff
@@ -640,8 +642,8 @@ def list_model_registrations(
     help='The number of GPUs used by the model, default is "auto".',
 )
 @click.option(
-    "--peft-model-path",
-    default=None,
+    "--peft-model-paths",
+    multiple=True,
     type=str,
     help="PEFT model path.",
 )
```
Comment (ChengjieLi28) [inline xinference/deploy/cmdline.py:649]: Note that this is cmdline, users cannot pass a List[str] directly. You need to state in the help message that it is comma-separated, and then split into List in the function.

File: xinference/deploy/cmdline.py [hunk 10]
```diff
@@ -696,7 +698,7 @@ def model_launch(
     quantization: str,
     replica: int,
     n_gpu: str,
-    peft_model_path: Optional[str],
+    peft_model_paths: Optional[List[str]],
```
Comment (ChengjieLi28) [inline xinference/deploy/cmdline.py:701]: This type is still `Optional[str]`

File: xinference/deploy/cmdline.py [hunk 11]
```diff
@@ -729,6 +731,10 @@ def model_launch(
         else None
     )
 
+    peft_model_config = PeftModelConfig(
+        peft_model_paths, image_lora_load_params, image_lora_fuse_params
+    )
+
```
Comment (ChengjieLi28) [inline xinference/deploy/cmdline.py:737]: Handle str split

File: xinference/model/core.py [hunk 12]
```diff
@@ -13,7 +13,9 @@
 # limitations under the License.
 
 from abc import ABC, abstractmethod
-from typing import Any, Dict, List, Optional, Tuple
+from typing import Any, List, Optional, Tuple
+
+from xinference.utils import PeftModelConfig
```
Comment (ChengjieLi28) [inline xinference/model/core.py:18]: Ditto

File: xinference/model/image/core.py [hunk 13]
```diff
@@ -17,6 +17,8 @@
 from collections import defaultdict
 from typing import Dict, List, Optional, Tuple
 
+from xinference.utils import PeftModelConfig
```
Comment (ChengjieLi28) [inline xinference/model/image/core.py:20]: Ditto

File: xinference/model/image/stable_diffusion/core.py [hunk 14]
```diff
@@ -45,20 +45,19 @@ def __init__(
...
+        if self._lora_model_paths is not None:
             logger.info(
                 f"Loading the LoRA with load kwargs: {self._lora_load_kwargs}, fuse kwargs: {self._lora_fuse_kwargs}."
             )
             assert self._model is not None
-            self._model.load_lora_weights(
-                self._lora_model_path, **self._lora_load_kwargs
-            )
+            for path in self._lora_model_paths:
+                self._model.load_lora_weights(path, **self._lora_load_kwargs)
```
Comment (ChengjieLi28) [inline xinference/model/image/stable_diffusion/core.py:60]: Did you try this? Is this correct? First load all lora weights, then just fuse_lora?

File: xinference/model/llm/core.py [hunk 15]
```diff
@@ -20,6 +20,8 @@
 from collections import defaultdict
 from typing import TYPE_CHECKING, Dict, List, Optional, Tuple, Union
 
+from xinference.utils import PeftModelConfig
```
Comment (ChengjieLi28) [inline xinference/model/llm/core.py:23]: Ditto

**Comment by ChengjieLi28** [comment]
Please add some screenshots and comments to explain what model and which lora(s) you used and what effect you achieved.

**Comment by ChengjieLi28** [comment]
Also, you need to modify the releated documents: https://inference.readthedocs.io/zh-cn/latest/models/lora.html

File: xinference/model/image/stable_diffusion/core.py (hunk 14 above)
Comment (hainaweiben) [reply xinference/model/image/stable_diffusion/core.py]: The model qwen14B, using both [peft_qwen_14b_chat_int4_xlsum](https://huggingface.co/p1atdev/peft_qwen_14b_chat_int4_xlsum) and [Qwen-14b-chat-32k-lora](https://huggingface.co/yuyijiong/Qwen-14b-chat-32k-lora), yields the following results:
image

File: xinference/client/restful/restful_client.py [hunk 16]
```diff
@@ -799,10 +800,8 @@ def launch_model(
         quantization: Optional[str] = None,
         replica: int = 1,
         n_gpu: Optional[Union[int, str]] = "auto",
+        peft_model_config: Optional[PeftModelConfig] = None,
```
Comment (ChengjieLi28) [inline xinference/client/restful/restful_client.py:803]: The entry of client does not import this new class `PeftModelConfig`. This breaks the usage. Just use dict, then convert to `PeftModelConfig` internally.

File: xinference/client/restful/restful_client.py [hunk 17]
```diff
@@ -799,10 +800,10 @@ def launch_model(
         quantization: Optional[str] = None,
         replica: int = 1,
         n_gpu: Optional[Union[int, str]] = "auto",
-        request_limits: Optional[int] = None,
-        peft_model_path: Optional[str] = None,
+        peft_model: Optional[List[Dict]] = None,
         image_lora_load_kwargs: Optional[Dict] = None,
         image_lora_fuse_kwargs: Optional[Dict] = None,
```
Comment (ChengjieLi28) [inline xinference/client/restful/restful_client.py:805]: Rename this to `peft_model_config`, same as the internal naming.
Also, I think you can delete `image_lora_load_kwargs` and `image_lora_fuse_kwargs` parameters and just use one `peft_model_config` parameter.
Corresponding changes to documentation and function comments.

**Comment by hainaweiben** [comment]
# Test for multi lora 
image
The model gemma-it-2B, using both [1rsh/gemma-2b-it-fr](https://huggingface.co/1rsh/gemma-2b-it-fr) and [alfredplpl/gemma-2b-it-ja-poc](https://huggingface.co/alfredplpl/gemma-2b-it-ja-poc), yields the following results:
image

**Review by ChengjieLi28 — APPROVED** [review]

**Pull request merged**